### ISSUE TITLE
Consider changing build systems

### ISSUE THREAD

**Issue by kirbyfan64** [kobalicek/asmjit#71]
Ok...

I still remember an issue where you complained that CMake's language never works like you'd expect. I hate having to download a 20-dozen-file program and building it just to use AsmJit.

It also seems like the CMake script has LOTS of boilerplate (for instance, the Asmjit_Add_* macros).

So...I was wondering if you'd look into using another build system such as [fbuild](https://github.com/felix-lang/fbuild), [Boost.Build](www.boost.org/build/), [Waf](https://code.google.com/p/waf/), [Gyp](https://code.google.com/p/gyp/), or maybe even *shudder* SCons *shudder*.

Thoughts? I have experience using all of the above except SCons (and somewhat Gyp), so I could help with whatever you need.

As a point of interest, Boost.Build has built-in support for selecting between a static and a shared library.

**Comment by kobalicek** [comment]
Well, I have designed asmjit in a way that no build system is really needed and people that don't like cmake can just include files in their project(s) and it will magically work.

The reason I'm using cmake is that it is somehow able to generate project files for all platforms that I use and it's still possible to use it in a way I do; for example I can embed asmjit in another project without even building a static library - I don't think that this is supported in other build systems as there is always a "product".

I would not use waf, gyp or scons as I have always experienced problems with these, and I don't really know the others you mentioned. I mean cmake just seems to be a lesser evil at the moment, but if there is a new build system, that is not using python, can generate project files, and is really cross-platform and solves all things in a much more elegant way cmake does, I would be really interested ;)

**Comment by kirbyfan64** [comment]
Boost.Build does NOT generate project files, but it doesn't use Python and is VERY much cross-platform, and it's really flexible (albeit poorly documented). It's primarily used to build Boost, which maked it obvious that it has strong cross-platform support.

Fbuild uses Python, but it's also truly cross-platform and is easily my favorite build tool.

[Tundra](https://github.com/deplinenoise/tundra) doesn't generate project files, either, but it doesn't use Python. I'm not quite sure too much else about this tool, but it seems very nice.

[Premake](http://industriousone.com/premake) looks like it might be your best bet. It generates project files for lots of IDEs and is very elegant.

**Comment by kobalicek** [comment]
Well, I think that cmake is simply the best from these, it's really widely supported and devs usually have it installed anyway. Premake was here before cmake and didn't make it, and I have never heard of Tundra for example.

I mean unless there is a really good reason I would not switch. I have several projects using cmake and although I don't like its syntax at all I found it to be the tool that actually _works_ and if there is a problem I'm probably not the first one facing it.

I think the main problem is that I'm not really a cmake expert and I'm making some things more complicated than they are, so the script itself is probably much more bloated than it should be :)

**Comment by kirbyfan64** [comment]
I guess that makes sense...

It just bugs me that CMake is huge, takes forever to build, and generates the longest, slowest makefiles on the planet (4000+ lines for a 1-source, 1-executable program) which for some stupid reason call back into CMake, along with the fact that the CodeBlocks generator just reverts back to make and doesn't set up executables correctly. The only non-suckish part is the Ninja generator.

That is also the longest sentence I have ever written...

I was using CMake for a project that uses AsmJit. Then it irritated me, so I hopped between 20 build systems before using fbuild. Since I didn't want Python to be a dependency, I added a Boost.Build script as an option. Both were very short and elegant and made me much, much happier. ;)

**Comment by kobalicek** [comment]
Well, I didn't really study the output, but yeah, that's a bit crazy.

I would even consider adding more build systems so people can pick what they consider the best, but I'm just afraid that I will be fixing build scripts instead of coding in such case :)

**Comment by kobalicek** [comment]
BTW I would also like to have a build system that works on arm, so maybe I will switch into something else in the future if cmake won't support it

**Comment by kirbyfan64** [comment]
What exactly do you mean by "natively"? Nothing really runs on Android itself except ports of GNU make.

I know that Boost.Build works on Android with [very minor tweaks](http://stackoverflow.com/a/17672227/2097780) that require nothing other than the NDK itself (the example in the link looks daunting, but it's really easier than it looks).

There are also CMake scripts for Android, but it is an extra dependency.

Of course, if you wanted *real* Android support, you could torture yourself AND AsmJit's users by using Gradle. :)

**Comment by kirbyfan64** [comment]
@kobalicek So...I was toying around and experimenting...and guess what? A fully-functional Boost.Build build system for is just 56 lines long! Gave me no trouble whatsoever. It's [here](https://gist.github.com/kirbyfan64/549151151b4a440c48b2). The only cheat was that I used globs to shorten the file. Even without the globs, it's over 200 lines shorter than the CMake build script and it actually a bit more portable.

It's also a tad easier to read. ;)

**Comment by kobalicek** [comment]
I meant to compile asmjit on the Android device itself, this is currently not supported by cmake and I'm not sure how about the other build systems, but not in a rust with this as there are odroids that I can use for ARM development ;)

And yeah, the build looks cool, but I'm still not convinced to switch, maybe I can add it as an alternative, but still I don't really feel like maintaining two build scripts if one already works :)

**Comment by kirbyfan64** [comment]
@kobalicek

Boost.Build *theoretically* should be able to be built to run on Android, assuming someone [patches the Boehm configure script](https://github.com/schani/clojurec/blob/master/run/android/README.md) to make it pick up the NDK correctly (stupid autotools).

But...I would've thought that CMake would run, considering Android now actually has the C++ standard library.

**Comment by kobalicek** [comment]
Well, if more asmjit users actually come here and convince me I will add this build script as an alternative to cmake, +boost is sufficient comment for me to know there is a need for it.

But anyway, I really think that asmjit is so simple project to build, there are basically no dependencies and configuration is controlled through compile-time macros, so in the end I think that everybody is just embedding anyway :)

**Comment by wjakob** [comment]
There are a vast number of C++ capable build systems for which the same could be said (i.e. prettier, fewer lines of code, faster): Google build, Scons, Meson, premake, Maiken, ... One could probably go on for quite a bit.

Still, my vote would be for the keeping the existing cmake build system. It's not pretty but it is the de-facto standard for many C++ projects, and it facilitates interoperability with other CMake-based projects that recursively include asmjit using the "add_subdirectory" or "ExternalProject_add" mechanisms.

**Comment by lukedodd** [comment]
CMake isn't great but it's pretty much won the cross platform C++ build wars for now. With CMake you fit "natively" into the highest number of projects build systems. That is: I can include asmjit as a submodule and add it to my own projects CMake with add_subdirectory and link to it.

Non CMake people (or just the lazy) can just include the cpp/h files in the project themselves which is hardly any effort.

If I see a python based build system for a project and I'm working on Windows I usually sigh/laugh and move on.

tl;dr: +1 for CMake from me.

**Comment by kirbyfan64** [comment]
Well, it seems like I've lost this battle...

**Comment by kobalicek** [comment]
Not lost, but atm cmake fits the most people although I don't really like it as well :)
